### DISCUSSION TITLE
Argon2

### DISCUSSION THREAD

**Discussion by dancerqueen** [dyne/tomb#521] [Q&A] [answered]
what is the exact command to type in terminal to enable argon2?

Hi, so sorry to ask a simple question to all of you but its not simple for me...
I want to turn on argon2 but i dont know how.  i see post #517 but i dont understand solution because i can enter command in terminal if im told what to type but i dont understand a lot of what im typing.

ALSO...is there somewhere that lists the commands to enter other optional features?  i want to try some of them but dont know where to find the detailed command to enter for them.

last question...is there a command to upgrade from 2.9 to 2.11?  im installing new linux distro and it has 2.9 in repo and i would like to install that and upgrade to 2.11 if possible.

thanks for helping me!  i'm trying to learn this stuff but im not so smart so it takes me longer to learn thins but im trying.

**Comment by Narrat** [comment]
> what is the exact command to type in terminal to enable argon2?

This is something one chooses when creating a keyfile for a tomb. So the command in question would look like:
~`tomb forge --kdf 3 --kdftype=argon2 nameof.key`~
Edit: `tomb forge --kdf 3 --kdftype argon2 nameof.key`
Somewhere in the output should the following line appear: `tomb (*) Using KDF Argon2`
(Beware: I choose the 3 randomly. You may choose a value to your own liking :) )
 
> Hi, so sorry to ask a simple question to all of you but its not simple for me... I want to turn on argon2 but i dont know how. i see post #517 but i dont understand solution because i can enter command in terminal if im told what to type but i dont understand a lot of what im typing.

The usual reminder. Don't blindly type what a random person wrote on the internet :) Before executing check the help output or the man/info pages to gather an understanding of what is being done.

> ALSO...is there somewhere that lists the commands to enter other optional features? i want to try some of them but dont know where to find the detailed command to enter for them.

Worthwhile to check for every program and not only tomb:
```
$ man <name of tool>
$ <name of tool> --help (or -h for shorthand)
```
With the help option you will usually get a concise list of possebilities and with the man page a more elaborate documentation. Which also allow to search for keywords. 
https://wiki.gentoo.org/wiki/Man_page/Navigate if the topic of man-pages is new for you :)  

> last question...is there a command to upgrade from 2.9 to 2.11? im installing new linux distro and it has 2.9 in repo and i would like to install that and upgrade to 2.11 if possible.

This is a more elaborate topic.
First question would be: What distro did you choose? Some of the distros allow rather easily to create packages which can be installed with the package manager.
The classic way would be the holy trinity of
```
$ ./configure
$ make
# make install
```
if a project has a `Makefile`. Tomb does: https://github.com/dyne/tomb/blob/master/Makefile
But then one must make sure  that all dependencies are installed. In case of tomb rather easy as it is a shell script.
And lastly one could adjust the `$PATH` env to create a folder for programs in `$HOME` where the new 2.11 tomb script could be placed and can take precedence over other installed tomb versions.

> thanks for helping me! i'm trying to learn this stuff but im not so smart so it takes me longer to learn thins but im trying.

You're welcome and ask away.

**Reply by dancerqueen** [reply]
thanks so much for using your time to help me!  I will try to learn quickly.....

So it seems like the argon2 is set up with each file at time its created, my misunderstanding was thinking it was set up 1 time somehow and would apply automatically to each file going forward.  No problem now.  I created a test file using your example and it worked fine.
Question1: What does the "3" value represent?  I see it can be something else but Im not sure what to choose.
Question2: Is there a way to look at a file and see if Argon2 was used?  I will have some that do and do not have it so it would be nice to be able to check at some time to see this if possible.

The linux distro i chose was the Devuan.  I was hoping it would be easy to upgrade from tomb 2.9 to 2.11 since both are from the dyne org...  If i need to install as new I will try that way if needed.  I read on the web that the following will install the needed apps for the tomb to run, is this correct?
sudo apt install gnupg zsh cryptsetup pinentry-curses
Before i was just using the version in the distro that was available but now I want to use the 2.11 version and the 2.9 version is what is available from within Devuan.


do you recommend any other options to use for security, i'm most interested in the most secure way to have the file.  seems like now it will be aes256 and the argon2 used if i understand correctly.  i will be storing the password on external device.  anything you can help me with is great.  i'm using this to protect all my important data like banking and doctor info.  I will be writing down all this in deetail for my family if something happens to me and i will try to teach this to them so they can also use to protect their info if they are interested.

thanks again for helping me get better with this!!!!!

**Reply by Narrat** [reply, marked as answer]
> thanks so much for using your time to help me! I will try to learn quickly.....
> 
> So it seems like the argon2 is set up with each file at time its created, my misunderstanding was thinking it was set up 1 time somehow and would apply automatically to each file going forward. No problem now. I created a test file using your example and it worked fine.

It is a one time operation in that sense that you create a keyfile to lock a tomb and this keyfile uses additional KDF to make the password of the keyfile more robust against bruce force attacks.
Some stuff to read on that topic ;)
https://en.wikipedia.org/wiki/Key_derivation_function
https://en.wikipedia.org/wiki/Argon2
https://en.wikipedia.org/wiki/PBKDF2
https://bitwarden.com/help/kdf-algorithms/
https://community.bitwarden.com/t/pbkdf2-vs-argon2-which-is-better/59187

> Question1: What does the "3" value represent? I see it can be something else but Im not sure what to choose. 

The number chosen there will represent the # of iterations. So 3 iterations from the example. (which is also the default for argon2)
Simply said iterations are how often it will reiterate the KDF part before it can be checked if the input is valid as a password.
Argon2 and pbkdf are different in that regard. With pbkdf you will see a high number of iterations (>10k), whereas argon2 has a default of 3. Difference comes through inner workings.

> Question2: Is there a way to look at a file and see if Argon2 was used? I will have some that do and do not have it so it would be nice to be able to check at some time to see this if possible.

Yes. The keyfile itself is human readable. So you can simply use any program which can display text files. On the terminal this job could be done with `cat` or `less`. (Just don't save any changes to that file ;) )
If kdf was used, this will be visible from the first line in the keyfile, as the algorithm used will be visible there.

> The linux distro i chose was the Devuan. I was hoping it would be easy to upgrade from tomb 2.9 to 2.11 since both are from the dyne org... If i need to install as new I will try that way if needed.

Not firm with the structure of dyne, but for me it is more of a governing body under which may likeminded projects can gather. Devuan itself is a project to remove systemd from Debian and therefore uses everything possible from the mother distribution to stay compatible in a way. And therefore the packages and versions are in general the same. Depending on what Debian version it is build upon this may take some time.
Debian itself offers the following:
* buster (older supported release): tomb 2.5
* bullseye (older supported release): tomb 2.9
* bookworm (current release): tomb 2.9
* trixi (next big release): tomb 2.10
* sid (rolling release): tomb 2.10

I assume Sid will be first to get 2.11, but that depends on the maintainer of said package.

> I read on the web that the following will install the needed apps for the tomb to run, is this correct?
> `sudo apt install gnupg zsh cryptsetup pinentry-curses`

That is the bare minimum needed, yes. To see every optional feature of tomb: `tomb --version`

> Before i was just using the version in the distro that was available but now I want to use the 2.11 version and the 2.9 version is what is available from within Devuan.

Then you either need to look into creating your own `.deb` package or install it via `make install` or put it into your $HOME directory.

> do you recommend any other options to use for security, i'm most interested in the most secure way to have the file. seems like now it will be aes256 and the argon2 used if i understand correctly. i will be storing the password on external device. anything you can help me with is great. i'm using this to protect all my important data like banking and doctor info. I will be writing down all this in deetail for my family if something happens to me and i will try to teach this to them so they can also use to protect their info if they are interested.

That is such a difficult and general question and really depends on what one wants to achieve. More technical steps can reduce comfort and therefore make it difficult for non tech person to get access. Although that can be somewhat circumvented, as tomb has some extra tools (like little GUI helpers).
Did you think of options to tweak for tomb or additional programs?
In general the defaults are good and if you keep the keyfile on a separate device.. a lot done for the safety of your sensible personal data.
But don't forget Backups. If the external device gets lost...
 
> thanks again for helping me get better with this!!!!!

Again you're welcome :)

**Reply by dancerqueen** [reply]
This is good info!!!  I feel better now but will spend more time reading the stuff you linked to.  I know this took you time to spend writing this for me and again I cant say thank you enough!!!

I going to play around with the argon2 for a while before I begin using it with important data and then I will create new folders with argon2 and move all the data stored in the existing files i have over to them.

I'm not so smart with the command line stuff but I figured out Tomb so far with the instructions on the web site and thanks to your help I'll be able to migrate to include the argon2, this would have taken me forever, if that was even possible, to figure this out myself.  once i get this understood better i will write down easy instructions for my family to access my data if something happens to me.  And yes, backups not forgotten...I learned this the hard way in the past and lost data because a hard drive was defective and all data got lost so now i make a few backup copies now.

Its people like you in the linux community that give me a reason to stay and keep learning and getting better because your nice and helpful to me with learning this type of stuff like with command line things.  So thanks again...

**Comment by jaromil** [comment]
Thanks for taking care of answering @Narrat . All questions make sense, so much that I think this is useful input for a FAQ, ping @Sakrecoer .

**Reply by dancerqueen** [reply]
Thanks for noticing these chats and suggesting a FAQ!

**Comment by dancerqueen** [comment]
@Narrat A bit more help please....

I entered:
tomb dig -s 100 secrets3.tomb
tomb forge --kdf 3 --kdftype=argon2 secrets3.tomb.key
tomb lock -k secrets3.tomb.key secrets3.tomb

There was no output stating "Using KDF Argon2" or even just Argon2 anywhere.

I was able to open/close the file but I cant figure out how to confirm Argon2 was used.
I tried to open the keyfile with a text editor but what appeared was the encrypted text with Begin PGP and ending with End PGP just like when a .asc file is opened.

The only thing I new to do was open with
gpg --list-packets and gpg -vv --show-session-key --list-packets
but neither provided anything about Argon2

Please provide me w/any details I can try to confirm Argon2 or even get Argon2 to appear in the output in case I entered the info wrong creating the file.

Sorry for the trouble, I've tried all i can think of before contacting you again....

**Reply by dancerqueen** [reply]
@Narrat  UPDATE
I just reviewed the output again and noticed the following: Cipher is "aes" mode "xts-plain64" hash "sha512"
Maybe I'm wrong but it seems like it is not Argon2 but I've been up all night working on this and maybe i'm just too tired to think correctly and I was only looking for the word Argon2.  Just thought this might help explain where i'm at with this....

**Reply by Narrat** [reply]
Regarding looking at the key. Here an example of a key specificily created for this (pbkdf in use):
```
_KDF_pbkdf2sha1_eba6f3f105286aece6a16e46a545533cbd299d844da0ffa17b26dc001196f064_3011504_64

-----BEGIN PGP MESSAGE-----

jA0ECQMC1oqAwZYfJEf50ukB5iwyUvOYO1HRXhmezTpr0+CULqRK9px+2xHG25dW
I+bHZSaJHeC6OrZstPURKCVW/PHtEEVB/WLSG2mfw9dE6FAqoHjWDM8d8CS91+Hk
9JpGQiMDZOAwQntUrDBvuAVkAhYj2Tu7dTr5Wlv58v/BKrZsVK0YV5PE9qQIKGsV
2gKrGRS6Nti5GG60SWlnfvsfGXw4XKpBC9oUi5CTIU3pUCwh35iLpbuGy1Eh5II4
XSIkpQRQALtKzZt83sl16sE7FIDeSYORdnSXMNVwd0wn4vc2iiHFpOycWQw/EyvR
g2mOuS84JhSApLDRkefoeuzDYC7arsdmohhTOVcuTNwoGF7d9HiV8xQSHzzAoB9S
sB88J8V1cOg9MZgD4r07d+t2YwIBqoJUcNOlNYems2mrW7BJYE05PsnGdyWR79GU
VQRk+MqnajBA7Knxtp6zFB82H/5/+f2UW2FrElRMrRHmz7ziPk4wt2kOzLICpNCe
1lNDuF4yZGEUjTQ8RYTnD67MRp5DP+OrsXOIE6/qHVDveMlE++h7JRas+Z2AxYIF
Ra3ir+yNjC4MQmaTIYLXCONdn6H1afDC0Bpmvz8QymSOvdcPcyEp8yB9axX8aqOW
hWz2lvOmPAZx/0SIvpjIHg9KUb3N9cwH45OGaaMa/36QMLii+zd7nUSpY3X6tJXK
7TfMPe53TUne+vQt9tJvf6cGc5UquTSfIIIuR1Jwyz4vtV/mKromSV3TTNTNuEKP
eWZuD7TqWbFy
=Zg1B
-----END PGP MESSAGE-----
```
Without pbkdf or argon2 in use while creating you only would see the `BEGIN` and `END PGP` messages. In this example you can see as the first line `_KDF_pbkdf2_` which tells, that pbkdf2 was used.
So if don't see such a line in the file, then this mechanic is not in place. And it is likely `argon2` wasn't used while creating your key. I didn't check back when you mentioned what tomb version you're using. `argon2` support was only available with the [2.10 release](https://github.com/dyne/tomb/blob/2b75962e54c3f29f546d6ab8b46e124e86592bd4/ChangeLog.md?plain=1#L17). So no luck with 2.9...

And not to confuse with the cipher. The cipher part of the encryption of the tomb. IIRC `aes` with `xts-plain64` is still the default of `cryptsetup` for `LUKS` container. Simply said `AES256` is in use, like you wished :)
https://wiki.gentoo.org/wiki/Dm-crypt_full_disk_encryption#Which_cipher:hash_combination.3F
https://security.stackexchange.com/questions/5158/for-luks-the-most-preferable-and-safest-cipher

**Reply by dancerqueen** [reply]
The screenshot helps me know for sure what to look for 4 confirmation!
I didnt get this so i know 4 sure i did something wrong.

So what i did was reinstall the OS so it is clean install and not any Tomb pieces installed from the v2.9.  I installed the v2.11 with also gnupg, zsh, cryptsetup, pinentry-curses, argon2.

Maybe i didnt install all that is needed 4 v2.11....Seems like i'm doing something to make it more difficult than it needs to be....

Bad news= lookin like i'll just have 2 wait until 2.10/2.11 gets added to the software repository & see if this works 4 me then instead of installing it myself, i was just excited to use the most recent version w/argon2.

Good news = you taught good info so i'll spend the rest of my time reading over the material/links you provided and maybe monitor the web to see if anyone else is having same problems.

Thanks again 4 being so nice 2 me and helping!

**Reply by Narrat** [reply]
> The screenshot helps me know for sure what to look for 4 confirmation! I didnt get this so i know 4 sure i did something wrong.
> 
> So what i did was reinstall the OS so it is clean install and not any Tomb pieces installed from the v2.9. I installed the v2.11 with also gnupg, zsh, cryptsetup, pinentry-curses, argon2.
> 
> Maybe i didnt install all that is needed 4 v2.11....Seems like i'm doing something to make it more difficult than it needs to be....

What did happen? Core requirements didn't change that much over the past versions.
Did you enter `tomb (...)` in the terminal and it couldn't be found? If it was generally available what didn't work? `tomb --version` may help to see what is still missing.
In case you didn't know: You can in most terminal emulators mark text with the mouse and copy it with `Ctrl-Shift-c`, just similar to GUI editors and the like. Just a different key-combo as `Ctrl-c` was already in use :D 

> Bad news= lookin like i'll just have 2 wait until 2.10/2.11 gets added to the software repository & see if this works 4 me then instead of installing it myself, i was just excited to use the most recent version w/argon2.

Don't give up ;) Best learning experience to figure out why something didn't work.
 
> Good news = you taught good info so i'll spend the rest of my time reading over the material/links you provided and maybe monitor the web to see if anyone else is having same problems.
> 
> Thanks again 4 being so nice 2 me and helping!

You're welcome

**Reply by dancerqueen** [reply]
OK, thanks i'll keep going, u r right, i was just tired & frustrated after a long weekend of trying to get this working....it has to be simpler than i'm making it & i'm curious for what i'm doin wrong....

THIS IS TOMB VERSION INFO


$ tomb --version
  Tomb 2.11.0 - a strong and gentle undertaker for your secrets

   Copyright (C) 2007-2024 Dyne.org Foundation, License GNU GPL v3+
   This is free software: you are free to change and redistribute it
   For the latest sourcecode go to <http://dyne.org/software/tomb>

   This source code is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
   When in need please refer to <http://dyne.org/support>.

  System utils:

  zsh 5.9 (x86_64-debian-linux-gnu)
  Sudo version 1.9.13p3
  cryptsetup 2.6.1 flags: UDEV BLKID KEYRING KERNEL_CAPI
  pinentry-gnome3 (pinentry) 1.2.1
  findmnt from util-linux 2.38.1
  gpg (GnuPG) 2.2.40 - key forging algorithms (GnuPG symmetric ciphers):
  IDEA 3DES CAST5 BLOWFISH AES AES192 AES256 TWOFISH CAMELLIA128 CAMELLIA192 CAMELLIA256

  Optional utils:

  /usr/bin/gettext
  dcfldd not found
  /usr/bin/shred
  steghide not found
  /sbin/resize2fs
  tomb-kdb-pbkdf2 not found
  /usr/bin/argon2
  qrencode not found
  recoll not found
  unoconv not found
  /usr/bin/lsof


i installed gnupg, zsh, cryptsetup, pinentry-curses, argon2






i did 2 tests so 1 without argon2 (file named "noargon2") and 1 with argon2 (file named "withargon2")

TEST1 i did
tomb dig -s 100 noargon2.tomb
tomb forge noargon2.tomb.key
tomb lock noargon2.tomb -k noargon2.tomb.key

all seems 2 b OK, it lets me open & close no problem.
keyfile is open & it doesnt have anything above the -----BEGIN PGP MESSAGE----- or after the -----END PGP MESSAGE-----
this seems 2 b correct 2 me if i understand correctly.



TEST2 i did
tomb dig -s 100 withargon2.tomb
tomb forge --kdf 3 --kdftype=argon2 withargon2.key     (i also tried using withargon2.tomb.key)
tomb lock withargon2.tomb -k withargon2.key

all "seems" 2 b OK, it lets me open and close no problem.
BUT....keyfile is open it *doesnt* have anything above the -----BEGIN PGP MESSAGE----- or after the -----END PGP MESSAGE-----, its same as the test1 with no argon2.
AND....i dont see argon2 in the output that i see, i pasted the info from terminal below for the withargon2 file


TERMINAL INFO

$ tomb dig -s 100 withargon2.tomb
tomb  .  Commanded to dig tomb
tomb (*) Creating a new tomb in withargon2.tomb
tomb  .  Generating withargon2.tomb of 100MiB
100+0 records in
100+0 records out
104857600 bytes (105 MB, 100 MiB) copied, 3.11118 s, 33.7 MB/s
-rw------- 1 100M withargon2.tomb
tomb (*) Done digging withargon2.tomb
tomb  .  Your tomb is not yet ready, you need to forge a key and lock it:
tomb  .  tomb forge withargon2.tomb.key
tomb  .  tomb lock withargon2.tomb -k withargon2.tomb.key
$ tomb forge --kdf 3 --kdftype=argon2 withargon2.key
tomb  .  Commanded to forge key withargon2.key with cipher algorithm AES256
tomb [W] This operation takes time. Keep using this computer on other tasks.
tomb [W] Once done you will be asked to choose a password for your tomb.
tomb [W] To make it faster you can move the mouse around.
tomb [W] If you are on a server, you can use an Entropy Generation Daemon.
512+0 records in
512+0 records out
512 bytes copied, 0.000564728 s, 907 kB/s
tomb (*) Choose the password of your key: withargon2.key
tomb  .  (You can also change it later using 'tomb passwd'.)
tomb  .  Key is valid.
tomb  .  Done forging withargon2.key
tomb (*) Your key is ready:
-rw------- 1 855 withargon2.key
$ tomb lock withargon2.tomb -k withargon2.key
tomb  .  File is not yet a tomb: withargon2.tomb

[sudo] Enter password for user to gain superuser privileges


tomb  .  Valid tomb file found: withargon2.tomb
tomb  .  Commanded to lock tomb withargon2.tomb
tomb  .  Checking if the tomb is empty (we never step on somebody else's bones).
tomb  .  Fine, this tomb seems empty.
tomb  .  Key is valid.
tomb  .  Locking using cipher: aes-xts-plain64
tomb  .  A password is required to use key withargon2.key
tomb  .  Password OK.
tomb (*) Locking withargon2.tomb with withargon2.key
tomb  .  Formatting Luks mapped device.
tomb  .  Formatting your Tomb with ext4 filesystem.
tomb  .  Done locking withargon2 using Luks dm-crypt aes-xts-plain64
tomb (*) Your tomb is ready in withargon2.tomb and secured with key withargon2.key
$ tomb open withargon2.tomb -k withargon2.key
tomb  .  Commanded to open tomb withargon2.tomb
tomb  .  Valid tomb file found: withargon2.tomb
tomb  .  Key is valid.
tomb  .  Mountpoint not specified, using default: /media/withargon2
tomb (*) Opening withargon2 on /media/withargon2
tomb  .  This tomb is a valid LUKS encrypted device.
tomb  .  Cipher is "aes" mode "xts-plain64" hash "sha512"
tomb  .  A password is required to use key withargon2.key
tomb  .  Password OK.
DM-UUID for device tomb.withargon2.53f73e2bab7280dbc32bfab58ffc8912f4c0c18b10477ca20052e1eb0e3ea8d4.loop0 was truncated.
tomb (*) Success unlocking tomb withargon2
tomb  .  Filesystem detected: ext4
tomb  .  Checking filesystem via /dev/loop0
fsck from util-linux 2.38.1
withargon2: clean, 11/25064 files, 11951/100352 blocks
tomb (*) Success opening withargon2.tomb on /media/withargon2
$ tomb close
tomb  .  Closing tomb [withargon2] mounted on /media/withargon2
tomb (*) Tomb [withargon2] closed: your bones will rest in peace.






Does this help?
More info i can give u?
Can u enter what i entered & see if you get same stuff?
do i need to install the tomb-kdb-pbkdf2 that says not found?  i didnt look into this because i think i remember the msg before said pbkdf2 or argon2 so i ignored since argon2 is installed.  if i need tomb-kdb-pbkdf2 installed also, what is the command to install this?

as usual, thanks!!!!!

**Reply by Narrat** [reply]
That helped tremendously :) Thanks
In general all is looking good. Essential deps are installed with argon2 being available. And in general tomb is working.
And it not necessary to install `tomb-kdf-pbkdf2`.
What happened is that I erred in one part, when I gave the example command:
`$ tomb --kdf 3 --kdftype=argon2 test.key`
The issue is with the `=` on `--kdftype`.

```
$ tomb forge --kdf 3 --kdftype=argon2  withargon3.key
tomb  .  Commanded to forge key withargon3.key with cipher algorithm AES256
tomb  .  Using KDF to protect the key password (3 rounds)
tomb [W] This operation takes time. Keep using this computer on other tasks.
tomb [W] Once done you will be asked to choose a password for your tomb.
tomb [W] To make it faster you can move the mouse around.
tomb [W] If you are on a server, you can use an Entropy Generation Daemon.
512+0 records in
512+0 records out
512 bytes copied, 0.00132025 s, 388 kB/s
tomb (*) Choose the password of your key: withargon3.key
tomb  .  (You can also change it later using 'tomb passwd'.)
tomb [W] Key is missing KDF header.
tomb [W] The key does not seem to be valid.
tomb [W] Dumping contents to screen:

tomb [W] 
tomb [E] Operation aborted.
```
Although it is odd, that it didn't fail for you. Maybe it tried to fall back to `pbkdf2` and as it wasn't installed it bailed.

So what is wrong? Often longform command line options which accept an argument can be written in two ways `--option=foo` and `--option foo`. This is not case here. The right command is without the `=`: `$ tomb --kdf 3 --kdftype argon2 test.key`

Output:
```
$ tomb forge --kdf 3 --kdftype argon2 withargon4.key 
tomb  .  Commanded to forge key withargon4.key with cipher algorithm AES256
tomb  .  Using KDF to protect the key password (3 rounds)
tomb [W] This operation takes time. Keep using this computer on other tasks.
tomb [W] Once done you will be asked to choose a password for your tomb.
tomb [W] To make it faster you can move the mouse around.
tomb [W] If you are on a server, you can use an Entropy Generation Daemon.
512+0 records in
512+0 records out
512 bytes copied, 0.0014562 s, 352 kB/s
tomb (*) Choose the password of your key: withargon4.key
tomb  .  (You can also change it later using 'tomb passwd'.)
tomb (*) Using KDF Argon2
tomb  .  memory used: 2^18
tomb  .  kdf salt: fb23da0b3c863b727c4c8a13f4ded14225f77787168500dd5a8bbbbaa44f784a
tomb  .  kdf iterations: 3
tomb  .  Key is valid.
tomb  .  Done forging withargon4.key
tomb (*) Your key is ready:
```
with this being the first lines in the resulting file:
```
_KDF_argon2_fb23da0b3c863b727c4c8a13f4ded14225f77787168500dd5a8bbbbaa44f784a_3_18_64

-----BEGIN PGP MESSAGE-----
(...)
```

So please try again the command without the `=` and I'm sorry for the confusion my wrong example created.

**Reply by dancerqueen** [reply]
Thanks!  But....still not working.....
I create new file name using withargon4 to be just like your example, here is terminal info

$ tomb forge --kdf 3 --kdftype argon2 withargon4.key
tomb  .  Commanded to forge key withargon4.key with cipher algorithm AES256
tomb [W] This operation takes time. Keep using this computer on other tasks.
tomb [W] Once done you will be asked to choose a password for your tomb.
tomb [W] To make it faster you can move the mouse around.
tomb [W] If you are on a server, you can use an Entropy Generation Daemon.
512+0 records in
512+0 records out
512 bytes copied, 0.000601293 s, 851 kB/s
tomb (*) Choose the password of your key: withargon4.key
tomb  .  (You can also change it later using 'tomb passwd'.)
tomb  .  Key is valid.
tomb  .  Done forging withargon4.key
tomb (*) Your key is ready:

any thing more i can do to help solve this puzzle?

**Reply by Narrat** [reply]
I should have been more curious on the fact, that you didn't get any warning whatsoever.
But I looked at the code to figure out what happens. And now I know more and imo the whole situation can be improved. Taking currently a stab at this.
The whole KDF handling code only checks only on `pbkdf2` and not `argon2`. So without `tomb-kdf-pbkdf2` installed, no KDF for your key. It is suboptimal, as this shouldn't be necessary. But on the other hand this is currently somewhat right, as the argon implementation in tomb uses `tomb-kbd-pbkdf2-gensalt` to generate a salt. And this is typically installed alongside `tomb-kdb-pbkdf2`

Long story short: You currently need the `pbkdf2` related tools.
Those can be found in the tomb source in the subfolder extras
https://github.com/dyne/tomb/tree/master/extras/kdf-keys
Via `make` the binaries can be compiled. And the resulting binaries can be installed via `make install` as root. Assuming you did the same with tomb. How did you install the newest tomb version?

**Reply by dancerqueen** [reply]
To install i downloaded the tar.gz file and did:
cd Downloads
tar xvfz Tomb-2.11.tar.gz
cd Tomb-2.11
sudo make install

i think this is all i did to install but i also installed the zsh, gnupg, cryptsetup, pinentry-curses, argon2.

if you tell me what to enter for this i will give it a try...

is this correct?  the link says
type make and need to have installed apt-get install libgcrypt11-dev

do i install  apt-get install libgcrypt11-dev and then just type the work "make"?
or is there more to enter?

thanks again!!

**Reply by dancerqueen** [reply]
@Narrat Update: the libgcrypt11-dev wasnt there but i got libgcrypt20-dev installed.  not sure about what 2 do with the make and make install so ill wait for reply from u so i dont mess things up enter wrong info.

seems like maybe i type this below but i dont know 4 sure....
tomb-kdb-${algo}
tomb-kdb-${algo}-gensalt
tomb-kdb-${algo}-getiter
tomb-kdb-hexencode
sudo make install

**Reply by Narrat** [reply]
`libgcrypt20-dev` is good. The README is in that regard outdated and needs an update.

And like you did with tomb itself the same applies for the kbd-keyes folder :) :
``` shell
$ cd Tomb-2.11/extras/kdf-keys
$ make
$ sudo make install
```

**Reply by dancerqueen** [reply]
I think you got it working 4 me!!!!!!!!!
how does this look below?

this appears before the -----BEGIN PGP MESSAGE-----
_KDF_argon2_239bedd4429d1ff9d3a3cf1b1d754978bfb0657f968cf2918331334b4ca64e87_3_18_64

this appears in the terminal output
$ tomb forge --kdf 3 --kdftype argon2 argontest.key
tomb  .  Commanded to forge key argontest.key with cipher algorithm AES256
tomb  .  Using KDF to protect the key password (3 rounds)
tomb [W] This operation takes time. Keep using this computer on other tasks.
tomb [W] Once done you will be asked to choose a password for your tomb.
tomb [W] To make it faster you can move the mouse around.
tomb [W] If you are on a server, you can use an Entropy Generation Daemon.
512+0 records in
512+0 records out
512 bytes copied, 0.00270948 s, 189 kB/s
tomb (*) Choose the password of your key: argontest.key
tomb  .  (You can also change it later using 'tomb passwd'.)
tomb (*) Using KDF Argon2
tomb  .  memory used: 2^18
tomb  .  kdf salt: 239bedd4429d1ff9d3a3cf1b1d754978bfb0657f968cf2918331334b4ca64e87
tomb  .  kdf iterations: 3
tomb  .  Key is valid.
tomb  .  Done forging argontest.key
tomb (*) Your key is ready:


the key file for this also looks different.  the color of this is orange with a image of a square with arrow inside pointed to the right.
the key file with no argon2 is green with image of a key in the middle.

does this all look good??????

**Reply by Narrat** [reply]
Yes, this looks good :)

**Reply by dancerqueen** [reply]
THANK YOU for stayin w/me & help me get this working.  Dont know many who do this for me.  I learn a lot from this and still gonna read the info you gave me to learn more.

1 last question please....
when doin the forge i been gettin this:
tomb [W] This poses a security risk.
tomb [W] You can deactivate all swap partitions using the command:
tomb [W]  swapoff -a
tomb [W] [#163] I may not detect plain swaps on an encrypted volume.
tomb [W] But if you want to proceed like this, use the -f (force) flag.
tomb [E] Operation aborted.

so i been doin:
$ sudo swapoff -a
and then do the forge.

is this the way to do this best for security?

**Reply by Narrat** [reply]
No problem. In the end it was worthwhile for both sides.
Regarding your question and what you're doing. Yes, running `swapoff` removes a potential security risk. Although depending on the distribution sometimes swap partitions are per default encrypted nowadays. The tomb output mentions an issue here on github on this specific topic: #163 
General short info on swap: https://wiki.archlinux.org/title/Swap
Simply said it is a store on the disk where RAM pages can be moved into. Needed if RAM needs are greater than available or suspend/hibernation on a Laptop is used. And therefore potential important and critical data could end up in the Swap.
Encrypted swap: https://wiki.archlinux.org/title/Dm-crypt/Swap_encryption
Is possible, but comes with disadvantages and cannot currently reliable detected by tomb. Therefore the advice to turn of swap space.
